unblob/pool.py

```diff
@@ -107,9 +150,17 @@ def submit(self, args):
 
     def process_until_done(self):
         while not self._input.is_empty():
-            result = self._output.get()
-            self._result_callback(self, result)
-            self._input.task_done()
+            try:
+                result = self._output.get()
+                if result is _SENTINEL:
+                    logger.debug("Shutting down pool, remaining tasks cancelled")
+                    break
+
+                self._result_callback(self, result)
+                self._input.task_done()
+            except ShutDownRequired:
+                logger.debug("Shutdown requested, remaining tasks will be cancelled")
+                self._output.put(_SENTINEL)
```

Comment:
I know that I came up with this design in the first place but it doesn't make much sense: We can just move the try block outside of the loop so that we can exit the loop without touching the output queue at all.